source/fab/engine.py

```diff
@@ -0,0 +1,190 @@
+                 target: str,
+                 pathmaps: List[PathMap],
+                 taskmap: Mapping[
+                     Tuple[Type[FileType], Type[State]],
+                     Task]) -> None:
+        self._workspace = workspace
+        self._target = target
+        self._pathmaps = pathmaps
+        self._taskmap = taskmap
+        self._database = SqliteStateDatabase(workspace)
+
+    @property
+    def target(self) -> str:
+        return self._target
+
+    def process(self,
+                artifact: Artifact,
+                discovery,
+                objects,
+                lock) -> List[Artifact]:
+
+        new_artifacts = []
+        new_discovery = {}
+        new_objects = []
+        # Identify tasks that are completely new
+        if (artifact.state is New
+                and artifact.filetype is Unknown):
+            # Use the pathmap list to work out the
+            # filetype and starting state
+            new_artifact = None
+            for pathmap in self._pathmaps:
+                if pathmap.match(artifact.location):
+                    new_artifact = Artifact(artifact.location,
+                                            pathmap.filetype,
+                                            pathmap.state)
+            # Assuming we found a match and were able
+            # to create the artifact, return it so that
+            # it can be added to the queue
+            if new_artifact is not None:
+                # Also store its hash in the file database
+                file_info = FileInfoDatabase(self._database)
+                file_info.add_file_info(artifact.location,
+                                        new_artifact.hash)
+                new_artifacts.append(new_artifact)
+
+        elif artifact.state is Analysed:
+
+            # Work out whether this artifact needs to be
+            # included in the build or not - tag it as having
+            # been seen so the system knows it is Analysed
+            required = False
+            for definition in artifact.defines:
+                if definition in discovery:
+                    new_discovery[definition] = "Seen"
+                    required = True
+
+            # Assuming it is needed, check its
+            # dependencies to know what needs doing
+            if required:
+                compiled = [False]*len(artifact.depends_on)
+                for idep, dependency in enumerate(artifact.depends_on):
+                    if dependency in discovery:
+                        # Are the dependencies compiled?
+                        if discovery[dependency] == "Compiled":
+                            compiled[idep] = True
+                    else:
+                        # If the dependency isn't in the list at all yet
+                        # then add an entry so the system knows we are
+                        # expecting it later (for the above check)
+                        new_discovery[dependency] = "HeardOf"
```

Comment:
This "discovery" dictionary may become an object in its own right with methods for updating the state of each unit. Another thought for the future.

Reply:
I played about with this a little bit - though I may be wrong about this I couldn't quite see how to get the `Manager` (from `multiprocessing`) to let me use a generic object for these shared/proxy entities.  I agree it would be nicer if we can work it out though